### PR TITLE
[Source Tooling] Refactoring action to convert if statement to switch

### PR DESCRIPTION
**Pull request by Regno**
Implement action to convert if statements from:
  ```
var e = E.case1
  if e == .case1 {}
  else if e == .case2 {}
  else {}
```
to
```
  var e = E.case1
  switch e {
  case .case1: break
  case .case2: break
  default: break
  }
```

Resolves [SR-5740](https://bugs.swift.org/browse/SR-5740).

@nkcsgexi

### CODE REVIEW

**Comment by Regno** [comment]
@nathawes Could you review this PR please?

**Review by nathawes — CHANGES_REQUESTED** [review]
Looking good so far! I thought of a few edge cases that might need handling and some tests to add while looking over it. I mentioned some potential enhancements to make it applicable in more situations in case you're interested in working on them, but those can certainly be left for a future PR too.

**Comment by Regno** [comment]
I can't reopen this PR after force push to branch. So I create new one: https://github.com/apple/swift/pull/29596

File: lib/IDE/Refactoring.cpp [hunk 1]
```diff
@@ -2243,6 +2243,104 @@ bool RefactoringActionConvertGuardExprToIfLetExpr::performChange() {
...
+      return false;
+    auto FirstConditionElement = CondList[0];
+    auto ConditionKind = FirstConditionElement.getKind();
+    if (ConditionKind != swift::StmtConditionElement::CK_Boolean)
+      return false;
+    auto ConditionExpr = dyn_cast<BinaryExpr>(FirstConditionElement.getBoolean());
+    if (!ConditionExpr)
+      return false;
+    auto ConditionArguments = ConditionExpr->getArg();
+    auto FirstArgument = dyn_cast<DeclRefExpr>(ConditionArguments->getElement(0));
```
[the remark this answers is not in the document]
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: I added test cases to check such situation:
```
func checkInverseCondition(e: E) {
   if .first == e {
     print("first")
   } else if .second == e {
     print("second")
   } else {
     print("default")
   }
 }
```
and
```
func checkMultipleOrConditions(e: E) {
   if e == .first || e == .second || e == .third {
     print("1-3")
   } else if e == .fourth { 
     print("4")
   } else { 
     print(">4")
   }
 }
```

https://github.com/apple/swift/pull/29596/files#diff-25cfa169a425e3d40c7c27e0135e3d58

File: lib/IDE/Refactoring.cpp [hunk 2]
```diff
@@ -2243,6 +2243,104 @@ bool RefactoringActionConvertGuardExprToIfLetExpr::performChange() {
...
+    if (CondList.size() != 1)
+      return false;
+    auto FirstConditionElement = CondList[0];
+    auto ConditionKind = FirstConditionElement.getKind();
+    if (ConditionKind != swift::StmtConditionElement::CK_Boolean)
+      return false;
+    auto ConditionExpr = dyn_cast<BinaryExpr>(FirstConditionElement.getBoolean());
+    if (!ConditionExpr)
+      return false;
+    auto ConditionArguments = ConditionExpr->getArg();
```
[the remark this answers is not in the document]
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: I added check to allowed functions names like `==` and `~=`.

File: test/refactoring/ConvertToSwitchStmt/basic.swift [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+func withoutElse(e: E) {
+  if e == .case1 {
+    print("case1")
+  } else if e == .case2 {
+    print("case2")
+  } else if e == .case3 {
+    print("case3")
+  }
+}
```
[the remark this answers is not in the document]
Comment (Regno) [reply test/refactoring/ConvertToSwitchStmt/basic.swift]: Added regression test:
```
func checkMultipleWithUnrelatedCondition(e: E, x: Int) {
   if e == .first || x > 0 {
     print("e is the first or x is positive")
   } else { 
     print("Shouldn't convert to switch")
   }
 }
```

File: lib/IDE/Refactoring.cpp [hunk 4]
```diff
@@ -2243,6 +2243,104 @@ bool RefactoringActionConvertGuardExprToIfLetExpr::performChange() {
...
+      auto FirstArgument = dyn_cast<DeclRefExpr>(ConditionArguments->getElement(0));
+      auto FirstArgumentValue = dyn_cast<VarDecl>(FirstArgument->getDecl());
+      OS << tok::kw_switch << Space << FirstArgumentValue->getNameStr().str() << Space << tok::l_brace << NewLine;
+    }
+    IsBegin = false;
+
+    auto SecondArgument = ConditionArguments->getElement(1);
+    auto SecondArgumentValue = Lexer::getCharSourceRangeFromSourceRange(SM, SecondArgument->getSourceRange());
+
+    auto IfBody = dyn_cast_or_null<BraceStmt>(If->getThenStmt());
```
[the remark this answers is not in the document]
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: Added test with empty `If` statement body:

```
func checkEmptyBody(e: E) {
   if e == .first {}
   else if e == .second {
     print("second")
   }
 }
```

File: lib/IDE/Refactoring.cpp [hunk 5]
```diff
@@ -2243,6 +2243,104 @@ bool RefactoringActionConvertGuardExprToIfLetExpr::performChange() {
...
+  do {
+    auto CondList = If->getCond();
+    auto FirstConditionElement = CondList[0];
+    auto ConditionExpr = dyn_cast<BinaryExpr>(FirstConditionElement.getBoolean());
+    auto ConditionArguments = ConditionExpr->getArg();
+
+    if (IsBegin) {
+      auto FirstArgument = dyn_cast<DeclRefExpr>(ConditionArguments->getElement(0));
+      auto FirstArgumentValue = dyn_cast<VarDecl>(FirstArgument->getDecl());
+      OS << tok::kw_switch << Space << FirstArgumentValue->getNameStr().str() << Space << tok::l_brace << NewLine;
```
[the remark this answers is not in the document]
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: Also added this case. Take a look at line 81 `checkLabeledCondition`.

File: lib/IDE/Refactoring.cpp [hunk 6]
```diff
@@ -2243,6 +2243,104 @@ bool RefactoringActionConvertGuardExprToIfLetExpr::performChange() {
...
+    auto ConditionExpr = dyn_cast<BinaryExpr>(FirstConditionElement.getBoolean());
+    if (!ConditionExpr)
+      return false;
+    auto ConditionArguments = ConditionExpr->getArg();
+    auto FirstArgument = dyn_cast<DeclRefExpr>(ConditionArguments->getElement(0));
+    auto FirstArgumentValue = dyn_cast<VarDecl>(FirstArgument->getDecl());
+    auto FirstArgumentType = FirstArgument->getType();
+    auto TypeKind = FirstArgumentType->getKind();
+    if (TypeKind != TypeKind::Enum)
+      return false;
```
[the remark this answers is not in the document]
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: I added such test cases too.

But I don't check that type conforms to the `Equatable` protocol. I just suppose that user will try to convert a valid `if` statement.

File: lib/IDE/Refactoring.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (nathawes) [reply lib/IDE/Refactoring.cpp]: The case I was thinking of was more if the user hasn't actually written the braces at all yet:
```
if e == .first || e == .second // stopped typing here to convert to switch
```
But no need to change it.

**Pull request closed without merging**